plotly_resampler/aggregation/plotly_aggregator_parser.py

```diff
@@ -0,0 +1,158 @@
+                # TODO: we can ommit pandas here, by using hasattr - see below
+                hf_data_dict[k] = hf_data_dict[k].values
+
+    @staticmethod
+    def to_same_tz(
+        ts: Union[pd.Timestamp, None], reference_tz
+    ) -> Union[pd.Timestamp, None]:
+        """Adjust `ts` its timezone to the `reference_tz`."""
+        if ts is None:
+            return None
+        elif reference_tz is not None:
+            if ts.tz is not None:
+                assert ts.tz.zone == reference_tz.zone
+                return ts
+            else:  # localize -> time remains the same
+                return ts.tz_localize(reference_tz)
+        elif reference_tz is None and ts.tz is not None:
+            return ts.tz_localize(None)
+        return ts
+
+    @staticmethod
+    def get_start_end_indices(hf_trace_data, start, end) -> Tuple[int, int]:
+        """Get the start & end indices of the high-frequency data."""
+        # Base case: no hf data, or both start & end are None
+        if not len(hf_trace_data["x"]):
+            return 0, 0
+        elif start is None and end is None:
+            return 0, len(hf_trace_data["x"])
+
+        # NOTE: as we use bisect right for the end index, we do not need to add a
+        #      small epsilon to the end value
+        start = hf_trace_data["x"][0] if start is None else start
+        end = hf_trace_data["x"][-1] if end is None else end
+
+        # We can compute the start & end indices directly when it is a RangeIndex
+        if isinstance(hf_trace_data["x"], pd.RangeIndex):
+            x_start = hf_trace_data["x"].start
+            x_step = hf_trace_data["x"].step
+            return max((start - x_start) // x_step, 0), (end - x_start) // x_step
+        # NOTE: this can be performed as-well for a fixed frequency range-index w/ freq
+
+        if hf_trace_data["axis_type"] == "date":
+            start, end = pd.to_datetime(start), pd.to_datetime(end)
+            # convert start & end to the same timezone
+            if isinstance(hf_trace_data["x"], pd.DatetimeIndex):
+                tz = hf_trace_data["x"].tz
+                assert start.tz == end.tz
+                start = PlotlyAggregatorParser.to_same_tz(start, tz)
+                end = PlotlyAggregatorParser.to_same_tz(end, tz)
+
+        # Search the index-positions
+        start_idx = bisect.bisect_left(hf_trace_data["x"], start)
+        end_idx = bisect.bisect_right(hf_trace_data["x"], end)
+        return start_idx, end_idx
+
+    @staticmethod
+    def aggregate(
+        hf_trace_data,
+        start_idx: int,
+        end_idx: int,
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+        """Aggregate the data in `hf_trace_data` between `start_idx` and `end_idx`.
+
+        Returns:
+            - x: the aggregated x-values
+            - y: the aggregated y-values
+            - indices: the indices of the hf_data data that were aggregated
```

Comment:
If the data is defined by a range, why are we creating a full index array? Let's explain how this might be useful.